new/language/implementation/javascript/javascript.go

```diff
@@ -0,0 +1,240 @@
+				return "", fmt.Errorf("invalid node type '%s' in pattern query", nodeType)
+			}
+		}
+
+		dummyValue := produceDummyValue(i, nodeTypes[0])
+
+		params = append(params, patternquerytypes.Variable{
+			Name:       submatches[nameIndex],
+			NodeTypes:  nodeTypes,
+			DummyValue: dummyValue,
+		})
+
+		i += 1
+
+		return dummyValue, nil
+	})
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	return replaced, params, nil
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) ExtractPatternMatchNode(input string) (string, int, error) {
+	inputBytes := []byte(input)
+	matches := matchNodeRegex.FindAllIndex(inputBytes, -1)
+
+	if len(matches) == 0 {
+		return input, 0, nil
+	}
+
+	if len(matches) > 1 {
+		return "", 0, errors.New("pattern must only contain a single match node")
+	}
+
+	match := matches[0]
+	return string(inputBytes[0:match[0]]) + string(inputBytes[match[1]:]), match[0], nil
+}
+
+// TODO: See if anything needs to be added here
+func produceDummyValue(i int, nodeType string) string {
+	switch nodeType {
+	case "identifier", "call":
+		return "curioVar" + fmt.Sprint(i)
+	default:
+		return "CurioVar" + fmt.Sprint(i)
+	}
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) AnonymousPatternNodeParentTypes() []string {
+	return anonymousPatternNodeParentTypes
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) FindPatternMatchNode(input []byte) [][]int {
+	return matchNodeRegex.FindAllIndex(input, -1)
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) FindPatternUnanchoredPoints(input []byte) [][]int {
+	return ellipsisRegex.FindAllIndex(input, -1)
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) IsTerminalDetectionNode(node *tree.Node) bool {
+	return false
+}
+
+// TODO: See if anything needs to be added here
+func (implementation *javascriptImplementation) PatternMatchNodeContainerTypes() []string {
+	return patternMatchNodeContainerTypes
+}
+
+func (implementation *javascriptImplementation) PatternIsAnchored(node *tree.Node) bool {
+	// FIXME: implement this
+	parent := node.Parent()
+	if parent == nil {
+		return true
+	}
+
+	// Class body
+	if parent.Type() == "class_declaration" && !node.Equal(parent.ChildByFieldName("name")) {
+		return false
+	}
+
+	// arrow functions
+	if parent.Type() == "arrow_functions" && !node.Equal(parent.ChildByFieldName("parameters")) {
+		return false
+	}
+
+	// function
+	if parent.Type() == "function_declaration" && !node.Equal(parent.ChildByFieldName("name")) && !node.Equal(parent.ChildByFieldName("parameters")) {
+		return false
+	}
+
+	// method
+	if parent.Type() == "method_definition" && !node.Equal(parent.ChildByFieldName("name")) && !node.Equal(parent.ChildByFieldName("parameters")) {
+		return false
+	}
+
+	return true
```

Comment:
I think this should be something like:
```suggestion
	return parent.Type() != "class_body" && parent.Type() != "statement_block"
```

Reply:
Saw it implemented like this in ruby and thought there is some reason for it.
will change it.

Reply:
In Ruby the expressions are directly under the class/block where there can be other things like the class name/block args. But in JS it's nicely in a separate node.